noria-server/src/controller/migrate/mod.rs

```diff
@@ -451,24 +542,93 @@ impl<'a> Migration<'a> {
                 dns
             });
 
-        // Boot up new domains (they'll ignore all updates for now)
-        debug!(log, "booting new domains");
-        for domain in changed_domains {
-            if mainline.domains.contains_key(&domain) {
-                // this is not a new domain
-                continue;
+        // Bucket reader replica nodes and the non-replica nodes separately.
+        let mut replica_map = HashMap::new();
+        let mut other_nodes = HashSet::new();
+        for &&ni in sorted_new
+                .iter()
+                .filter(|&&&ni| ni != mainline.source)
+                .filter(|&&&ni| !mainline.ingredients[ni].is_dropped())
+                .into_iter() {
+            // Check if the node is a reader, and then insert it in the replica map. The value
+            // is a list of readers that are for the same node, and the key is the node the
+            // readers are for. Later, we will also add the associated ingress node to the replica
+            // map, since readers are all located on their own domains.
+            if let Some(for_node) = mainline.ingredients[ni]
+                    .with_reader(|r| Some(r.is_for()))
+                    .unwrap_or(None) {
+                if mainline.ingredients[ni].replica_index().is_some() {
+                    replica_map.entry(for_node).or_insert(HashSet::new());
+                    replica_map.get_mut(&for_node).unwrap().insert(ni);
+                }
+            } else {
+                other_nodes.insert(ni);
             }
+        }
 
-            let nodes = uninformed_domain_nodes.remove(&domain).unwrap();
-            let d = mainline.place_domain(
-                domain,
-                mainline.ingredients[nodes[0].0].sharded_by().shards(),
-                &log,
-                nodes,
+        // Obtain a vec of DomainIndexes in the order that we want to assign them to workers.
+        // For replicas, all replicas for the same reader should be consecutive. For non-replicas,
+        // it doesn't really matter unless we want the ordering to be deterministic.
```

Comment:
This code below is pretty complicated, and it's not entirely clear why or how it does what it does. Could you try to simplify it?

Reply:
That helps a little, though I'm still not sure why we need to create *two* `changed_domains_*` and then call `place_round_robin` twice, as opposed to just having `place_round_robin` know about what things are readers and what things are not, and assign appropriately in a single pass?

Reply:
I could assign domains to replicas more like how it's done for shards, but I'm not actually sure how domain assignment works for shards -- seems like it's ok for shards to share domains with other nodes?

Since we're placing round robin, we have to specifically list readers for the same node in consecutive order in `changed_domain_readers`, to ensure their respective domains are placed on different workers. To only call `place_round_robin` once, we _could_ combine `changed_domain_*` so long as the readers are still in the right order, or move the reader-distinguishing logic into `place_round_robin`. Did you have something specific in mind?

Reply:
For shards, the logic is pretty straightforward, since any domain that is sharded simply gets "placed" as many times as there are shards ([here](https://github.com/mit-pdos/noria/pull/113/files#diff-c2efd5de7d5bd366a1b53929183e8c12R374)). I guess for replicas, this is harder since you don't have the same "one thing is many things" to rely on in `place`. I think it isn't so much that I think the approach you've taken is wrong or weird, it's just that I didn't fully understand it when I first read the code. It could be that a comment at the top about why we need to invoke `place_round_robin` twice would be enough.